### PR TITLE
Fix compatibility hash spoofing when plugins are present.

### LINKED ISSUE #58: When using NML alongside other plugins, NML should not ruin the compatibility hash.
The compatibility hash is normally altered for each loaded plugin. Imagine a function `hash` that takes a list of plugins. For vanilla, you'd have `hash()`. For NML, you'd have `hash(nml)`. For someone running multiple plugins, you'd have `hash(myCoolPlugin, otherPlugin, veryGoodPlugin)`.

Currently, if NML is the only loaded plugin it spoofs the hash to be `hash()`, enabling multiplayer compatibility with vanilla users. However, if there is another plugin being loaded as well, then NML completely disables its spoofing and uses `hash(NML, otherPlugin)`. This means that a otherPlugin user *not* running NML will have `hash(otherPlugin)` and therefore be incompatible with the NML + otherPlugin user.

NML should fix this by removing itself from the hash, but leaving any other plugins.

### PR DESCRIPTION
**Pull request by zkxs**
Resolves #58.

This enumerates all loaded assemblies and uses a couple of heuristics to make a *very* educated guess about which ones are plugins. Those plugins are then used in the compatibility hash calculation.

---

edit: I'll go into more detail

1. Enumerate all loaded assemblies using Froox's own `[module: Description("POSTX_PROCESSED")]` marker to filter to just the PostX processed assemblies
2. Attempt to correlate the PostX'd assemblies with Neos's `Engine.ExtraAssemblies`. ExtraAssemblies is a list of the loaded plugins, but it only contains their filenames. No path information.
3. Do some set theory to check for assemblies we couldn't match to plugins and plugins we couldn't match to assemblies.
   - assemblies we couldn't match to plugins aren't a huge deal, and we actually expect a few of these. Specifically, NeosModLoader, FrooxEngine, and BusinessX are all PostX'd assemblies we shouldn't treat as plugins. Anything not in this expected set I warn about.
   -  plugins we couldn't match to assemblies are a bigger deal, and indicate that my plugin detection code failed. We can't version spoof properly in this case.
4. Deal with the Unsafe and AdvertiseVersion configs
5. Make sure I sort the assemblies in the same order Froox does
6. Spoof the version string and the compatibility hash, but this time respecting plugins

### CODE REVIEW

**Comment by zkxs** [comment]
```
/home/runner/work/NeosModLoader/NeosModLoader/NeosModLoader/AssemblyLoader.cs(51,36): error WHITESPACE: Fix whitespace formatting. Delete 1 characters. [/home/runner/work/NeosModLoader/NeosModLoader/NeosModLoader/NeosModLoader.csproj]
/home/runner/work/NeosModLoader/NeosModLoader/NeosModLoader/Util.cs(77,1): error WHITESPACE: Fix whitespace formatting. Replace 17 characters with '\n\s\s\s\s\s\s\s\s'. [/home/runner/work/NeosModLoader/NeosModLoader/NeosModLoader/NeosModLoader.csproj]
/home/runner/work/NeosModLoader/NeosModLoader/NeosModLoader/Util.cs(78,69): error WHITESPACE: Fix whitespace formatting. Replace 9 characters with '\n\s\s\s\s\s\s\s\s'. [/home/runner/work/NeosModLoader/NeosModLoader/NeosModLoader/NeosModLoader.csproj]
```

I have no idea what the linter is complaining about. Those line/column targets don't appear valid to me.

**Comment by zkxs** [comment]
Some example logs using a couple of plugins I have lying around.

with `unsafe` false

```log
11:12:13 AM.760 (  0 FPS)	[DEBUG][NeosModLoader] Found 5 potential plugins: BusinessX.dll, FrooxEngine.dll, NeosDynamicBoneWizard.dll, NeosModLoader.dll, NQuery.dll
11:12:13 AM.768 (  0 FPS)	[DEBUG][NeosModLoader] Found 2 actual plugins: NeosDynamicBoneWizard.dll, NQuery.dll
11:12:13 AM.769 (  0 FPS)	[INFO] [NeosModLoader] Adding NeosModLoader.dll to version string because you have AdvertiseVersion set to true.
11:12:13 AM.771 (  0 FPS)	[DEBUG][NeosModLoader] Changing version string from 2022.1.28.1310+NeosModLoader.dll+NeosDynamicBoneWizard.dll+NQuery.dll to 2022.1.28.1310+NeosModLoader.dll+NeosDynamicBoneWizard.dll+NQuery.dll
11:12:13 AM.930 (  0 FPS)	[DEBUG][NeosModLoader] Vanilla protocol version is 843
11:12:13 AM.931 (  0 FPS)	[DEBUG][NeosModLoader] Changing compatibility hash from kTRVbxu53sCD7StS8J0Ehw== to rmJkVwWyIByK8lE5qvz+eA==
11:12:13 AM.931 (  0 FPS)	[INFO] [NeosModLoader] Compatibility hash spoofing succeeded
```

with `unsafe` true

```log
11:13:09 AM.257 (  0 FPS)	[DEBUG][NeosModLoader] Found 5 potential plugins: BusinessX.dll, FrooxEngine.dll, NeosDynamicBoneWizard.dll, NeosModLoader.dll, NQuery.dll
11:13:09 AM.264 (  0 FPS)	[DEBUG][NeosModLoader] Found 2 actual plugins: NeosDynamicBoneWizard.dll, NQuery.dll
11:13:09 AM.265 (  0 FPS)	[WARN] [NeosModLoader] Unsafe mode is enabled! Not that you had a warranty, but now it's DOUBLE void!
11:13:09 AM.265 (  0 FPS)	[INFO] [NeosModLoader] Adding NeosModLoader.dll to version string because you have AdvertiseVersion set to true.
11:13:09 AM.267 (  0 FPS)	[DEBUG][NeosModLoader] Changing version string from 2022.1.28.1310+NeosModLoader.dll+NeosDynamicBoneWizard.dll+NQuery.dll to 2022.1.28.1310+NeosModLoader.dll+NeosDynamicBoneWizard.dll+NQuery.dll
11:13:09 AM.420 (  0 FPS)	[DEBUG][NeosModLoader] Vanilla protocol version is 843
11:13:09 AM.421 (  0 FPS)	[DEBUG][NeosModLoader] Changing compatibility hash from kTRVbxu53sCD7StS8J0Ehw== to jnnkdwkBqGv5+jlf1u/k7A==
11:13:09 AM.421 (  0 FPS)	[INFO] [NeosModLoader] Compatibility hash spoofing succeeded
```

**Comment by ljoonal** [comment]
This sounds like a very hacky way to proceed, especially if we're having to deal with things such as 

> Make sure I sort the assemblies in the same order Froox does

...But I guess we can't do transpiler magic to just remove NeosModLoader, since the hash is calculated before NML loads?

**Comment by zkxs** [comment]
> This sounds like a very hacky way to proceed, especially if we're having to deal with things such as
> 
> > Make sure I sort the assemblies in the same order Froox does
> 
> ...But I guess we can't do transpiler magic to just remove NeosModLoader, since the hash is calculated before NML loads?

Correct, the compatibility hash is calculated before NML is loaded, and Neos does not keep a list of loaded plugins I can use to recalculate after the fact. So the real problem here is how to get the list of plugins. I have two ideas for this I've listed in https://github.com/neos-modding-group/NeosModLoader/issues/58#issuecomment-1218133308, and this PR uses the idea that, in my opinion, is the lesser of the evils.

The only way to get a non-hacky solution for this would for NML to not be a plugin at all, and to hook into Neos using a more "normal" modloader technique. And at that point we might as well just use a prexisting modloader, like BepinEx.

File: NeosModLoader/NeosVersionReset.cs [hunk 1]
```diff
@@ -12,65 +15,162 @@ namespace NeosModLoader
 {
 	internal class NeosVersionReset
 	{
+		// used when AdvertiseVersion == true
+		private const string NEOS_MOD_LOADER = "NeosModLoader.dll";
```
Comment (EIA485) [inline NeosModLoader/NeosVersionReset.cs:19]: users may change the name of their mod loader binary. we could use the filename of `Assembly.GetExecutingAssembly()`

File: NeosModLoader/NeosVersionReset.cs [hunk 2]
```diff
@@ -12,65 +15,162 @@ namespace NeosModLoader
...
 
 			if (!nmlPresent)
 			{
 				throw new Exception($"Assertion failed: Engine.ExtraAssemblies did not contain \"{assemblyFilename}\"");
 			}
 
-			bool otherPluginsPresent = extraAssemblies.Count > 1;
-			bool shouldSpoofCompatibility = !otherPluginsPresent || config.Unsafe;
-			bool shouldSpoofVersion = !config.AdvertiseVersion && shouldSpoofCompatibility;
+			// get all PostX'd assemblies. This is useful, as plugins can't NOT be PostX'd.
```
Comment (EIA485) [inline NeosModLoader/NeosVersionReset.cs:38]: some plugin authors use external tools to pre postX their plugins

Comment (zkxs) [reply NeosModLoader/NeosVersionReset.cs:38]: It's either PostX'd or not, and it doesn't matter if it's been done a month ago or seconds ago by neos.

File: NeosModLoader/NeosVersionReset.cs (hunk 1 above)
Comment (zkxs) [reply NeosModLoader/NeosVersionReset.cs:19]: My thoughts are that even if the DLL is named something silly like NeosModLoaderHeadless.dll I have a unique opportunity here to call it by its true name.

Also, this is only used if a user is using the `advertiseversion` config, which I sorta doubt many people do.

File: NeosModLoader/ModLoader.cs [hunk 3]
```diff
@@ -10,7 +10,7 @@ namespace NeosModLoader
 {
 	public class ModLoader
 	{
-		internal const string VERSION_CONSTANT = "1.11.3";
+		internal const string VERSION_CONSTANT = "1.12.0";
```
Comment (ljoonal) [inline NeosModLoader/ModLoader.cs:13]: I think that version bumps should usually just be done by @zkxs directly to the main branch in a commit that also tags it as the v1.12.0 for example.

**Review by ljoonal — APPROVED** [review]
Can't say I thoroughly reviewed this, but as it seemed okay and can always be changed later on, 👍

**Comment by EIA485** [comment]
> Can't say I thoroughly reviewed this, but as it seemed okay and can always be changed later on, 👍
yep, we can always change it later.

File: NeosModLoader/ModLoader.cs (hunk 3 above)
Comment (zkxs) [reply NeosModLoader/ModLoader.cs:13]: Yeah, you're probably right. It'd definitely make the commit log easier to read. I'll start doing that in the future.

**Pull request merged**